docs/proposals/event-notification-api.md

```diff
@@ -0,0 +1,119 @@
+
+This proposal does not include an implementation of a specific event notification provider for Origin. The scope of this proposal is limited to the minimal API necessary to enable the proposed features above.
+
+## Design
+
+An `Event` notifier integrating with Origin requires user-provided data to inform the notifier about interested users within a namespace and provide context for notification dispatch. Events are scoped to a namespace, and many users can access a namespace. Each `Event` could result in multiple notifications depending on the number of interested users and their notification configuration.
+
+For the proposed API, an `Event` type is uniquely identified by a key computed as `{Event.InvolvedObject.Kind}:{Event.Reason}`.
+
+### API
+
+The Event Notification API provides:
+
+1. A way for cluster operators to declare arbitrary notification types.
+2. A way for cluster operators to declare arbitrary notification constraints per event type.
+3. A way for users to declare their notification preferences per event type.
+
+```go
+// EventNotificationType is an event notification type supported generally the cluster.
+type EventNotificationType struct {
+  unversioned.TypeMeta
+  kapi.ObjectMeta
+  // Description is the human readable description of the type.
+  Description string
+}
+
+// ClusterEventNotification specifies which event notification types are supported
+// for a given event type within the cluster. There should only be one ClusterEventNotification
+// document in the cluster.
+type ClusterEventNotification struct {
+  unversioned.TypeMeta
+  kapi.ObjectMeta
+
+  // SupportedNotifications maps event types to supported notification types.
+  SupportedNotifications map[string][]string
+}
+
+// UserEventNotification specifies how a user should be notified of a given event.
+type UserEventNotification struct {
+  unversioned.TypeMeta
+  // Name is a User.Name.
+  kapi.ObjectMeta
+  // Notifications is an event type mapped to the names of EventNotificationTypes
+  // which are desired for the event.
+  Notifications map[string][]string
+}
+```
+
+### Console
+
+The Origin console can provide a management view for user notification settings based on the cluster configuration, dynamically generating a UI using the available metadata. If the cluster has no event notification types, the feature can be disabled entirely in the console or presented in a way that indicates the cluster doesn't support any notifications.
+
+### CLI
+
+TODO: Write me.
+
+## Use case: Cluster implements email and text message notifications
+
+The cluster operator supports dispatching emails and text messages to users via a third-party system. The integration is provided by a custom application deployed in Origin in a system namespace. The cluster operator wants to make email notifications available to users when `Pod:PodFailed` events occur, and make both email and text notifications available for `DeploymentConfig:Failed` events.
```

Comment:
I believe `Pod:Failed` is sufficient for a unique event reason.

The combination of Reason, Source, etc. is what makes the Event object itself distinct for compression.